### PR TITLE
sql: update `explain` results

### PR DESCRIPTION
**Pull request by LittleFall**
### What is changed, added or deleted? (Required)

Update statement of `EXPLAIN` with tidb-in-action.

Update explain result with current result.

### Which TiDB version(s) do your changes apply to? (Required)

master

### What is the related PR or file link(s)?

- [tidb pr/15507](https://github.com/pingcap/tidb/pull/15507)
- [tidb-in-action pr/424](https://github.com/pingcap-incubator/tidb-in-action/pull/424)

### CODE REVIEW

**Comment by TomShawn** [comment]
@LittleFall Please sign the CLA and involve a technical review, thanks!

**Comment by LittleFall** [comment]
> 要在这个 PR 中把整个阅读执行计划部分都用 《TiDB in Action》 中的内容更新一下吗？对了，`explain for connection` 部分也需要介绍，最好也在 《TiDB in Action》中把这部分内容加上。

整合了一下，PTAL

**Comment by LittleFall** [comment]
> The content in `reference/performance/understanding-the-query-execution-plan.md` is too long. We'd better to split it into several small sections.
> 
> Maybe we can organize the content like this:
> 
> ```
> file 1: EXPLAIN Basics, covered in https://pingcap.com/docs-cn/stable/reference/sql/statements/explain/#explain
> file 2: EXPLAIN FOR Connection, covered in https://pingcap.com/docs-cn/stable/reference/sql/statements/explain-analyze/
> file 3: EXPLAIN ANALYZE Extension, not covered in https://pingcap.com/docs-cn/stable/reference/sql/statements/explain/, this doc need to be updated.
> file 4: How to read the output of `EXPLAIN`, should be covered in this file
> file 5: How to read the output of `EXPLAIN ANALYZE`, maybe we can use another file to cover this topic.
> ```

有几个问题
1. 因为 understanding-the-query-execution-plan.md 中把算子信息都介绍过了，file5 中实际没什么可写的了。
2. 还有是不是 EXPLAIN FOR Connection 写在 explain.md 中比写在 explain-analyze.md 中要好一些？

调整了一下这些文件的结构，可能会比较清晰并且最长的文件看起来短一点。

explain.md:
1. 语法图
2. EXPLAIN 输出格式
3. 示例
4. MySQL 兼容性
5. `EXPLAIN FOR CONNECTION`

explain-analyze.md:
1. 语法图
2. EXPLAIN ANALYZE 输出格式
3. 示例
4. MySQL 兼容性

understanding-the-query-execution-plan.md
1. 使用 `EXPLAIN` 来优化 SQL 语句
2. 如何阅读算子的执行顺序
3. 算子的共有信息
4. 算子的独特信息
5. 优化实例

PTAL

**Comment by zz-jason** [comment]
@LittleFall 可以，这种格式会合理很多。我们需要尽量确保每个独立的文档阅读时间不超过 10 分钟。

File: reference/performance/understanding-the-query-execution-plan.md [hunk 1]
```diff
@@ -12,146 +11,366 @@ TiDB 优化器会根据当前数据表的实际情况来选择最优的执行计
...
+针对同一列的范围查询条件使用 `AND` 和 `OR` 组合后，等于对范围求交集或者并集。对于多维组合索引，可以写多个列的条件。例如对组合索引`(a, b, c)`，当 a 为等值查询时，可以继续求 b 的查询范围，当 b 也为等值查询时，可以继续求 c 的查询范围，反之如果 a 为非等值查询，则只能求 a 的范围。
+
+## 算子的独特信息
+
+### 如何阅读扫表的执行计划
+
+真正执行扫表（读盘或者读 TiKV Block Cache）操作的算子有如下几类：
+
+- **TableFullScan**：这是大家所熟知的 “全表扫” 操作
+- **TableRangeScan**：带有范围的表数据扫描操作，通常扫描的数据量不大
```
Comment (winoros) [inline reference/performance/understanding-the-query-execution-plan.md:93]: `通常扫描的数据量不大` 这句可能不需要保留？

File: reference/performance/understanding-the-query-execution-plan.md [hunk 2]
```diff
@@ -12,146 +11,366 @@ TiDB 优化器会根据当前数据表的实际情况来选择最优的执行计
...
++---------------------------+----------+-----------+---------------+---------------------------------+
+| HashAgg_11                | 1.00     | root      |               | funcs:count(Column#7)->Column#4 |
+| └─TableReader_12          | 1.00     | root      |               | data:HashAgg_5                  |
+|   └─HashAgg_5             | 1.00     | cop[tikv] |               | funcs:count(1)->Column#7        |
+|     └─TableFullScan_8     | 10000.00 | cop[tikv] | table:t       | keep order:false, stats:pseudo  |
++---------------------------+----------+-----------+---------------+---------------------------------+
+4 rows in set (0.00 sec)
 ```
 
+一般而言 TiDB 的 `Hash Aggregate` 会分成两个阶段执行，一个在 TiKV/TiFlash 的 `Coprocessor` 上，计算聚合函数的中间结果。另一个在 TiDB 层，汇总所有 `Coprocessor` Task 的中间结果后，得到最终结果。
```
Comment (winoros) [inline reference/performance/understanding-the-query-execution-plan.md:179]: 分成两个阶段执行的情况可以写清楚是 儿子是 TiKV/TiFlash 读取数据的算子的时候

File: reference/performance/understanding-the-query-execution-plan.md [hunk 3]
```diff
@@ -12,146 +11,366 @@ TiDB 优化器会根据当前数据表的实际情况来选择最优的执行计
...
+```
 
-在 WHERE/HAVING/ON 条件中，TiDB 优化器会分析主键或索引键的查询返回。如数字、日期类型的比较符，如大于、小于、等于以及大于等于、小于等于，字符类型的 LIKE 符号等。
+`Merge Join` 算子在执行时，会从 Build 端把一个 Join Group 的数据全部读取到内存中，接着再去读 Probe 端的数据，用 Probe 端的每行数据去和 Build 端的完整的一个 Join Group 依次去看是否匹配（除了满足等值条件以外，还有其他非等值条件，这里的 “匹配” 主要是指查看是否满足非等职条件）。Join Group 指的是所有 Join Key 上值相同的数据。
 
-值得注意的是，TiDB 目前只支持比较符一端是列，另一端是常量，或可以计算成某一常量的情况，类似 `year(birth_day) < 1992` 的查询条件是不能利用索引的。还要注意应尽可能使用同一类型进行比较，以避免引入额外的 cast 操作而导致不能利用索引，如 `user_id = 123456`，如果 `user_id` 是字符串，需要将 `123456` 也写成字符串常量的形式。
+**Index Hash Join 示例：**
 
-针对同一列的范围查询条件使用 `AND` 和 `OR` 组合后，等于对范围求交集或者并集。对于多维组合索引，可以写多个列的条件。例如对组合索引`(a, b, c)`，当 a 为等值查询时，可以继续求 b 的查询范围，当 b 也为等值查询时，可以继续求 c 的查询范围，反之如果 a 为非等值查询，则只能求 a 的范围。
+INL_HASH_JOIN(t1_name [, tl_name]) 提示优化器使用 Index Nested Loop Hash Join 算法。该算法与 Index Nested Loop Join 使用条件完全一样，但在某些场景下会更为节省内存资源。
```
Comment (winoros) [inline reference/performance/understanding-the-query-execution-plan.md:258]: 可以和 Hint 的文档也建立链接指向的关系

File: reference/performance/understanding-the-query-execution-plan.md [hunk 4]
```diff
@@ -12,146 +11,366 @@ TiDB 优化器会根据当前数据表的实际情况来选择最优的执行计
...
+|   └─TableRowIDScan_21(Probe)     | 1.00     | cop[tikv] | table:t1                 | keep order:false                                                         |
++----------------------------------+----------+-----------+--------------------------+--------------------------------------------------------------------------+
+8 rows in set (0.00 sec)
+```
 
-## Operator Info
+**Index Merge Join 示例：**
 
-### TableReader 和 TableScan
+INL_MERGE_JOIN(t1_name [, tl_name]) 提示优化器使用 Index Nested Loop Merge Join 算法。该算法相比于 INL_JOIN 会更节省内存。该算法使用条件包含 INL_JOIN 的所有使用条件，但还需要添加一条：join keys 中的内表列集合是内表使用的 index 的前缀，或内表使用的 index 是 join keys 中的内表列集合的前缀。
```
Comment (winoros) [inline reference/performance/understanding-the-query-execution-plan.md:279]: ditto

File: reference/performance/understanding-the-query-execution-plan.md [hunk 5]
```diff
@@ -12,146 +11,366 @@ TiDB 优化器会根据当前数据表的实际情况来选择最优的执行计
...
 
-TableScan 表示在 KV 端对表数据进行扫描，TableReader 表示在 TiDB 端从 TiKV 端读取，属于同一功能的两个算子。table 表示 SQL 语句中的表名，如果表名被重命名，则显示重命名。range 表示扫描的数据范围，如果在查询中不指定 WHERE/HAVING/ON 条件，则会选择全表扫描，如果在 int 类型的主键上有范围查询条件，会选择范围查询。keep order 表示 table scan 是否按顺序返回。
+```
+mysql> explain select /*+ INL_MERGE_JOIN(t1) */ * from t t1 where  t1.a  in ( select t2.a from t2 where t2.b < t1.b);
++------------------------------+----------+-----------+---------------+------------------------------------------------------------------------------------------------------+
+| id                           | estRows  | task      | access object | operator info                                                                                        |
++------------------------------+----------+-----------+---------------+------------------------------------------------------------------------------------------------------+
+| HashJoin_26                  | 8000.00  | root      |               | semi join, inner:TableReader_49, equal:[eq(test.t.a, test.t2.a)], other cond:lt(test.t2.b, test.t.b) |
```
Comment (lzmhhh123) [inline reference/performance/understanding-the-query-execution-plan.md:286]: 例子失效？

File: reference/performance/understanding-the-query-execution-plan.md [hunk 6]
```diff
@@ -12,146 +11,366 @@ TiDB 优化器会根据当前数据表的实际情况来选择最优的执行计
...
 
-Apply 是 TiDB 用来描述子查询的一种算子，行为类似于 Nested Loop，即每次从外表中取一条数据，带入到内表的关联列中，并执行，最后根据 Apply 内联的 Join 算法进行连接计算。
+## 另请参阅
 
-值得注意的是，Apply 一般会被查询优化器自动转换为 Join 操作。用户在编写 SQL 的过程中应尽量避免 Apply 算子的出现。
+* [EXPLAIN](/reference/sql/statements/explain.md)
+* [EXPLAIN ANALYZE](/reference/sql/statements/explain-analyze.md)
+* [ANALYZE TABLE](/reference/sql/statements/analyze-table.md)
+* [TRACE](/reference/sql/statements/trace.md)
+* [TiDB In Action](https://book.tidb.io/session3/chapter1/sql-execution-plan.html)
```
Comment (ran-huang) [inline reference/performance/understanding-the-query-execution-plan.md:376]:
```suggestion
* [TiDB in Action](https://book.tidb.io/session3/chapter1/sql-execution-plan.html)
```

File: reference/sql/statements/explain-analyze.md [hunk 7]
```diff
@@ -22,6 +22,17 @@ category: reference
 
 ![ExplainableStmt](/media/sqlgram/ExplainableStmt.png)
 
+## EXPLAIN ANALYZE 输出格式
+
+和 `EXPLAIN` 不同，`EXPLAIN ANALYZE` 会执行对应的 `SQL` 语句，记录其运行时信息，和执行计划一并返回出来，可以视为 `EXPLAIN` 语句的扩展。`EXPLAIN ANALYZE` 语句的返回结果相比 `EXPLAIN` 增加了 `actRows`, `execution info`,`memory`,`disk` 这几列信息：
```
Comment (ran-huang) [inline reference/sql/statements/explain-analyze.md:27]: Is this what you mean?
```suggestion
和 `EXPLAIN` 不同，`EXPLAIN ANALYZE` 会执行对应的 SQL 语句，记录其运行时信息，和执行计划一并返回出来。因此，可以将 `EXPLAIN ANALYZE` 视为 `EXPLAIN` 语句的扩展。`EXPLAIN ANALYZE` 语句的返回结果相比 `EXPLAIN`，增加了 `actRows`，`execution info`，`memory`，`disk` 这几列信息：
```

File: reference/sql/statements/explain.md [hunk 8]
```diff
@@ -24,6 +26,18 @@ category: reference
 
 ![ExplainableStmt](/media/sqlgram/ExplainableStmt.png)
 
+## EXPLAIN 输出格式
+
+目前 `TiDB` 的 `EXPLAIN` 会输出 5 列，分别是：`id`，`estRows`，`task`，`access object`， `operator info`。执行计划中每个算子都由这 5 列属性来描述，`EXPLAIN`结果中每一行描述一个算子。每个属性的具体含义如下：
+
+| 属性名          | 含义 |
+|:----------------|:----------------------------------------------------------------------------------------------------------|
+| id            | 算子的 ID，在整个执行计划中唯一的标识一个算子。在 TiDB 2.1 中，ID 会格式化的显示算子的树状结构。数据从孩子结点流向父亲结点，每个算子的父亲结点有且仅有一个。                                                                                       |
```
Comment (ran-huang) [inline reference/sql/statements/explain.md:35]: The spaces do not affect the eventual display on the PingCAP website.
```suggestion
| id            | 算子的 ID，是算子在整个执行计划中唯一的标识。在 TiDB 2.1 中，ID 会格式化地显示算子的树状结构。数据从孩子结点流向父亲结点，每个算子的父亲结点有且仅有一个。|
```

File: reference/sql/statements/explain.md [hunk 9]
```diff
@@ -24,6 +26,18 @@ category: reference
...
+| estRows       | 算子预计将会输出的数据条数，基于统计信息以及算子的执行逻辑估算而来。在 4.0 之前叫 count。 |
+| task          | 算子属于的 task 种类。目前的执行计划分成为两种 task，一种叫 **root** task，在 tidb-server 上执行，一种叫 **cop** task，并行的在 TiKV 或者 TiFlash 上执行。当前的执行计划在 task 级别的拓扑关系是一个 root task 后面可以跟许多 cop task，root task 使用 cop task 的输出结果作为输入。cop task 中执行的也即是 TiDB 下推到 TiKV 或者 TiFlash 上的任务，每个 cop task 分散在 TiKV 或者 TiFlash 集群中，由多个进程共同执行。 |
```
Comment (ran-huang) [inline reference/sql/statements/explain.md:37]:
```suggestion
| task          | 算子属于的 task 种类。目前的执行计划分成为两种 task，一种叫 **root** task，在 tidb-server 上执行，一种叫 **cop** task，在 TiKV 或者 TiFlash 上并行执行。当前的执行计划在 task 级别的拓扑关系是一个 root task 后面可以跟许多 cop task，root task 使用 cop task 的输出结果作为输入。cop task 中执行的也即是 TiDB 下推到 TiKV 或者 TiFlash 上的任务，每个 cop task 分散在 TiKV 或者 TiFlash 集群中，由多个进程共同执行。 |
```

**Comment by LittleFall** [comment]
I just adjust the structure of `performance/understanding-the-query-execution-plan.md`, like this:

```markdown
# 理解 TiDB 执行计划
## EXPLAIN 简介
### 算子的执行顺序
### Task 简介
### 范围查询
## 算子信息
### 如何阅读扫表的执行计划
#### 表数据和索引数据
#### IndexLookUp 示例
#### TableReader 示例
#### IndexMerge 示例
### 如何阅读聚合的执行计划
#### Hash Aggregate 示例
#### Stream Aggregate 示例
### 如何阅读 Join 的执行计划
.....(five join type)
## 优化实例
## 另请参阅
```

Thanks for your suggestions, PTAL.

**Comment by zz-jason** [comment]
```
## EXPLAIN 简介
### 算子的执行顺序
### Task 简介
### 范围查询
```

这里改成下面这样会更好些：

```
## EXPLAIN 简介
### Task 简介
### Task 的执行顺序
### Task 中算子的执行顺序
```

File: reference/sql/statements/explain.md [hunk 10]
```diff
@@ -24,6 +26,18 @@ category: reference
 
 ![ExplainableStmt](/media/sqlgram/ExplainableStmt.png)
 
+## EXPLAIN 输出格式
+
+目前 `TiDB` 的 `EXPLAIN` 会输出 5 列，分别是：`id`，`estRows`，`task`，`access object`， `operator info`。执行计划中每个算子都由这 5 列属性来描述，`EXPLAIN`结果中每一行描述一个算子。每个属性的具体含义如下：
```
Comment (ran-huang) [inline reference/sql/statements/explain.md:31]:
```suggestion
目前 TiDB 的 `EXPLAIN` 会输出 5 列，分别是：`id`，`estRows`，`task`，`access object`， `operator info`。执行计划中每个算子都由这 5 列属性来描述，`EXPLAIN`结果中每一行描述一个算子。每个属性的具体含义如下：
```

**Comment by zz-jason** [comment]
@LittleFall please resolve the code conflict, thanks.

File: reference/performance/understanding-the-query-execution-plan.md [hunk 11]
```diff
@@ -7,151 +7,372 @@ category: reference
...
 
-{{< copyable "sql" >}}
+IndexMerge 使得数据库在扫描表数据时可以使用多个索引。这里 `IndexMerge_16` 算子有三个孩子节点，其中 `IndexRangeScan_13` 和 `IndexRangeScan_14` 根据范围扫描得到符合条件的所有 RowID，再由 `TableRowIDScan_15` 算子根据这些 RowID 精确地读取所有满足条件的数据。
 
-```sql
-EXPLAIN SELECT count(*) FROM trips WHERE start_date BETWEEN '2017-07-01 00:00:00' AND '2017-07-01 23:59:59';
+> **注意：**
+>
+> 目前 TiDB 的 IndexMerge 特性在 4.0 RC 版本中默认关闭，同时 4.0 中的 IndexMerge 目前支持的场景仅限于析取范式（or 连接的表达式），对合取范式（and 连接的表达式）将在之后的版本中支持。
+> 开启 IndexMerge 特性，可通过在客户端中设置 session 或者 global 变量完成：`set @@tidb_enable_index_merge = 1;`
```
Comment (toutdesuite) [inline reference/performance/understanding-the-query-execution-plan.md:164]: @LittleFall 这里“目前...中支持”要与“开启 IndexMerge 特性这一句中间空一行，不然两行实际在显示时显示为一行~

Comment (LittleFall) [reply reference/performance/understanding-the-query-execution-plan.md:164]: 学习了，我之后一块改一下

**Pull request merged**